### PR TITLE
feat: calls protocol

### PR DESCRIPTION
**Pull request by lotharking**

### CODE REVIEW

File: src/events/MessageEvents.ts [hunk 1]
```diff
@@ -116,6 +126,56 @@ export const messageEvents = async (agent: ServiceAgent, config: ServerConfig) =
...
+
+      await sendMessageReceivedEvent(agent, msg, msg.timestamp, config)
+    }
+
+    if (message.type === CallEndMessage.type.messageTypeUri) {
+      const thread = (message as CallEndMessage).thread
+      const msg = new CallEndRequestMessage({
+        id: message.id,
+        connectionId: connection.id,
+        threadId: thread?.parentThreadId,
```
Comment (genaris) [inline src/events/MessageEvents.ts:147]:
```
        threadId: thread?.id,
```

File: src/events/MessageEvents.ts [hunk 2]
```diff
@@ -116,6 +126,56 @@ export const messageEvents = async (agent: ServiceAgent, config: ServerConfig) =
       await sendMessageReceivedEvent(agent, msg, msg.timestamp, config)
     }
 
+    if (message.type === CallOfferMessage.type.messageTypeUri) {
+      const parameters = (message as CallOfferMessage).parameters
+      const msg = new CallOfferRequestMessage({
+        id: message.id,
+        connectionId: connection.id,
+        requestedCallItem: parameters as unknown as RequestedCallItem,
+        threadId: message.thread?.parentThreadId,
```
Comment (genaris) [inline src/events/MessageEvents.ts:135]:
```
        threadId: message.thread?.id,
```

File: src/events/MessageEvents.ts [hunk 3]
```diff
@@ -116,6 +126,56 @@ export const messageEvents = async (agent: ServiceAgent, config: ServerConfig) =
...
+
+      await sendMessageReceivedEvent(agent, msg, msg.timestamp, config)
+    }
+
+    if (message.type === CallRejectMessage.type.messageTypeUri) {
+      const thread = (message as CallEndMessage).thread
+      const msg = new CallRejectRequestMessage({
+        id: message.id,
+        connectionId: connection.id,
+        threadId: thread?.parentThreadId,
```
Comment (genaris) [inline src/events/MessageEvents.ts:172]:
```suggestion
        threadId: thread?.id,
```

**Review by genaris — CHANGES_REQUESTED** [review]

File: src/controllers/message/MessageService.ts [hunk 4]
```diff
@@ -303,6 +305,28 @@ export class MessageService {
         await agent.connections.hangup({ connectionId: connection.id })
 
         // FIXME: No message id is returned here
+      } else if (messageType === CallOfferRequestMessage.type) {
+        const msg = JsonTransformer.fromJSON(message, CallOfferRequestMessage)
+        const { requestedCallItem } = msg
+
+        const callOffer = await agent.modules.calls.offer({
+          connectionId: connection.id,
+          callType: 'video',
```
[the remark this answers is not in the document]
Comment (lotharking) [reply src/controllers/message/MessageService.ts]: fixed

File: src/model/messages/MessageType.ts [hunk 5]
```diff
@@ -1,4 +1,8 @@
 export enum MessageType {
+  CallAcceptRequestMessage = 'call-accept-submit',
+  CallRejectRequestMessage = 'call-reject-submit',
+  CallEndRequestMessage = 'call-end-submit',
+  CallOfferRequestMessage = 'call-offer-request',
```
[the remark this answers is not in the document]
Comment (lotharking) [reply src/model/messages/MessageType.ts]: fixed

File: src/model/messages/calls/CallParameters.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  wsUrl: string
+  iceserver: IceserverItemOptions[]
+}
+
+export interface IceserverItemOptions {
+  urls: string
+  username: string
+  credential: string
+}
+
```
[the remark this answers is not in the document]
Comment (lotharking) [reply src/model/messages/calls/CallParameters.ts]: fixed

File: src/events/MessageEvents.ts (hunk 1 above)
Comment (lotharking) [reply src/events/MessageEvents.ts]: fixed, used to `thread?.threadId`

File: src/events/MessageEvents.ts (hunk 3 above)
Comment (lotharking) [reply src/events/MessageEvents.ts]: fixed, used to thread?.threadId

**Review by genaris — APPROVED** [review]

**Pull request merged**